mssql improvements, code clean up and fix issue 428

nmummau opened the pull request:
Each commit has isolated changes to a SQL table or procedure with detailed changes
No functional changes; these changes were a result of me doing a deep review of the procedure/tables that Eventuous creates.

This fixes #428

Other changes do include preferences, but I feel will aid to this SQL being easier to maintain for the long term. 
 <div id='description'>
    <a href="https://bito.ai#summarystart"></a>
<h3>Summary by Bito</h3>
This pull request addresses issue #428 by implementing improvements across PostgreSQL and SQL Server components. The PostgreSQL update corrects stream data ordering, while SQL Server changes refactor schema definitions and stored procedures. Updates standardize data types, constraints, and error handling to enhance transaction safety, maintainability, and performance. These modifications improve the overall reliability and consistency of the event store implementation.
<!-- Disabling unit_tests and post_effort_to_review
<br>
<br>
<b>Unit tests added</b>: False
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 3
-->
</div>

nealmummau committed:
feat(mssql): Minor SQL updates
- consistent casing of [Version] column
- explicitly specify NULLable columns

nealmummau committed:
feat(mssql): Minor SQL changes; no functional changes
- SET NOCOUNT ON
- remove unused declared variables
- specify all parameters when calling [check_stream] procedure
- formatting: each column on it's own line. Best practice for maintainability of procedures
- proper casing of keywords (IF, ISNULL, NVARCHAR)
- use alias to UPDATE statement
- remove 'AS' keyword
- semicolons

nealmummau committed:
feat(mssql): Minor SQL changes; no functional changes
- SET NOCOUNT ON
- formatting: each column on it's own line. Best practice for maintainability of procedures
- proper casing of keywords (AND)
- semicolons

nealmummau committed:
feat(mssql): Minor SQL changes; no functional changes
- SET NOCOUNT ON
- formatting: each column on it's own line. Best practice for maintainability of procedures
- use aliases
- semicolons

nealmummau committed:
feat(mssql): Minor SQL changes; no functional changes
- SET NOCOUNT ON
- Wrap keywords in brackets ([Version], [Messages])
- formatting: each column on it's own line. Best practice for maintainability of procedures
- proper casing of keywords (INT)
- remove 'Messages.' from query since it was redundant
- semicolons

nealmummau committed:
feat(mssql): Minor SQL changes; no functional changes
- SET NOCOUNT ON
- Wrap keywords in brackets ([Version], [Messages])
- formatting: each column on it's own line. Best practice for maintainability of procedures
- proper casing of keywords (INT)
- remove 'Messages.' from query since it was redundant
- semicolons

nealmummau committed:
feat(mssql): Minor SQL changes; no functional changes
- SET NOCOUNT ON
- Wrap keywords in brackets ([Messages])
- formatting: each column on it's own line. Best practice for maintainability of procedures
- proper casing of keywords (INT, BIGINT)
- remove 'Messages.' from query since it was redundant
- semicolons

nealmummau committed:
feat(mssql): Minor SQL changes; no functional changes
- SET NOCOUNT ON
- Wrap keywords in brackets ([Version], [Messages])
- formatting: each column on it's own line. Best practice for maintainability of procedures
- proper casing of keywords (INT, AND)
- use alias to DELETE statement
- semicolons

alexeyzimarev commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/1_Schema.sql:
```diff
@@ -25,9 +25,9 @@ IF OBJECT_ID('__schema__.Messages', 'U') IS NULL
-            Created        DATETIME2,
+            JsonData       NVARCHAR(MAX)         NOT NULL,
+            JsonMetadata   NVARCHAR(MAX)             NULL,
+            Created        DATETIME2                 NULL,
```
This column should not be nullable

nmummau replied:
Oh, interesting. Since these columns were not set they will be NULL, which is why I made this change just to be explicit with the definition.

If these columns should be NOT NULL, I can make that change but that should be communicated to anyone who already has the MSSQL as their event store. The recommendation would be to alter these columns setting them to NOT NULL.

nmummau replied:
@alexeyzimarev I agree that `Created` is safe to make NOT NULL since `append_events` will set that value if a NULL is provided to the procedure.

However, are you certain that `JsonMetadata` should be NOT NULL? I noticed it's nullable in both the `TYPE StreamMessage` and `record NewPersistedEvent`.

```sql
# 1_Schema.sql
CREATE TYPE __schema__.StreamMessage AS TABLE
(
    message_id    UNIQUEIDENTIFIER NOT NULL,
    message_type  NVARCHAR(128)    NOT NULL,
    json_data     NVARCHAR(MAX)    NOT NULL,
    json_metadata NVARCHAR(MAX)        NULL
)
```
```csharp
// SqlEventStoreBase.cs
public record NewPersistedEvent(Guid MessageId, string MessageType, string JsonData, string? JsonMetadata);
```

alexeyzimarev replied:
It can be nullable but I believe Eventuous adds an empty JSON (`{}`) if metadata is not provided by the user.

nmummau replied:
In that case I feel like the right thing to do is set the table definition to be NOT NULL. I personally am not setting any metadata but I'm using Otel so the span/trace are set in this field. I can do some testing today and determine if you are correct about that and update this PR accordingly.

nmummau replied:
In my testing, I'm getting data like the following in the Messages.JsonMetadata column
```
{"trace-id":"5ad9e899f7fd9542dbe86fcd12e4fa6e",
"span-id":"295a17e81ba88f66",
"parent-span-id":"13e8febe81d0989d"}
```

The MassTransit Worker service that I'm implementing Eventuous into already has Otel enabled, and to be honest I'm not entirely sure how this JsonMetadata is being set. I'm guessing that my service which is calling the CommandService is somehow getting this set?

Nonetheless, it seems that the right thing to do here is to set the JSON fields to NOT NULL. I've also added a CHECK constraint to the columns to ensure ISJSON is true.

alexeyzimarev replied:
Eventuous propagates upstream trace and follows it with its own spans. Check the diagnostics section in the docs, it's quite detailed about traces in particular. Weird that the parent span ID is being set, I thought I removed it.

nmummau committed:
fix(mssql): Messages.Created NOT NULL

SQL Server schema change so that Messages.Created is NOT NULL. This is safe and accurate to do since append_events will always set this value if a NULL is provided to the procedure.

nealmummau committed:
feat(mssql): json is not null

- Messages TABLE
    - JsonMetaData is NOT NULL
    - json columns are checked to ensure they contain json data
- StreamMessage TYPE
    - json_metadata is NOT NULL

nealmummau committed:
feat(mssql): DATETIME2 is actually a DATETIME2(7) so explicitly say that in the scripts

alexeyzimarev commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/6_ReadStreamForwards.sql:
```diff
@@ -4,24 +4,39 @@ CREATE OR ALTER PROCEDURE __schema__.read_stream_forwards
+    FROM __schema__.[Messages]
+    WHERE StreamId = @stream_id
+    AND StreamPosition >= @from_position
+    ORDER BY GlobalPosition;
```
GlobalPosition order is also incremental. But, indeed it would be clearer if the order is by StreamPosition

nmummau replied:
@alexeyzimarev  would you like me to apply that change?

nealmummau committed:
feat(mssql): ORDER BY StreamPosition per bito-code-review, and and remove redundencies in the CHECK definition

nealmummau committed:
feat(postgresql): ORDER BY stream_position per bito-code-review

nmummau commented:
/review

nmummau commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/1_Schema.sql:
```diff
@@ -25,14 +25,16 @@ IF OBJECT_ID('__schema__.Messages', 'U') IS NULL
-            Created        DATETIME2,
+            JsonData       NVARCHAR(MAX)         NOT NULL,
+            JsonMetadata   NVARCHAR(MAX)         NOT NULL,
+            Created        DATETIME2(7)          NOT NULL,
```
The `append_events` procedure already handles this, being sure that a value is set.

nmummau commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/1_Schema.sql:
```diff
@@ -53,8 +55,7 @@ IF TYPE_ID('__schema__.StreamMessage') IS NULL
             message_type  NVARCHAR(128)    NOT NULL,
-            json_data     NVARCHAR(max)    NOT NULL,
-            json_metadata NVARCHAR(max)
+            json_data     NVARCHAR(MAX)    NOT NULL,
+            json_metadata NVARCHAR(MAX)    NOT NULL
```
This comment is wrong. Both the TYPE and TABLE are both NOT NULL now

nmummau replied:
Good catch! I amended a commit and fixed this.

nealmummau committed:
refactor(mssql): remove square brackets to satisfy 'inconsistent table naming' warning that bito-code-review flagged

nealmummau committed:
feat(mssql): append_events improve

At first when writing this change I assumed I would be out to fix issue #429. Turns out that is not an issue as the client wraps this procedure call in a transaction. Thus, no transction is needed within the procedure itself. Here are some improvements, nonetheless.

- Changed @stream_name parameter to NVARCHAR(850) for consistency
- Introduced @inserted table to capture inserted GlobalPosition values atomically
- Normalized optimistic concurrency errors (2627, 2601) to client-detectable messages
- Thrown error messages still StartsWith 'WrongExpectedVersion' but now also includes the full SQL error text for client tracing
- Removed redundant post-commit queries and error-parsing logic

nealmummau committed:
perf(mssql): enable OPTIMIZE_FOR_SEQUENTIAL_KEY on primary keys for Streams and Messages

- Added WITH (OPTIMIZE_FOR_SEQUENTIAL_KEY = ON) to clustered primary keys on __schema__.Streams (PK_Streams) and __schema__.Messages (PK_Events)
- Improves insert throughput and reduces latch contention on identity-based keys
- No changes to schema shape, constraints, or indexes beyond PK optimization

This improves concurrent insert performance under high write workloads by allowing SQL Server to better handle last-page insert contention on sequential identity keys.
See docs on OPTIMIZE_FOR_SEQUENTIAL_KEY: https://learn.microsoft.com/en-us/sql/t-sql/statements/alter-table-index-option-transact-sql?view=sql-server-ver17#optimize_for_sequential_key---on--off-

nmummau commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/2_AppendEvents.sql:
```diff
@@ -1,50 +1,78 @@
             BEGIN
-                DECLARE @streamIdFromError nvarchar(20) = SUBSTRING(ERROR_MESSAGE(), PATINDEX(N'%[0-9]%,%', ERROR_MESSAGE()), PATINDEX(N'%, [0-9]%).', ERROR_MESSAGE()) - PATINDEX(N'%[0-9]%,%', ERROR_MESSAGE()))
-                DECLARE @streamPositionFromError nvarchar(20) = SUBSTRING(ERROR_MESSAGE(), (PATINDEX(N'%, [0-9]%).', ERROR_MESSAGE())) + 2, PATINDEX(N'%).', ERROR_MESSAGE()) - (PATINDEX(N'%, [0-9]%).', ERROR_MESSAGE()) + 2))
+                DECLARE @streamIdFromError NVARCHAR(20) = SUBSTRING(ERROR_MESSAGE(), PATINDEX(N'%[0-9]%,%', ERROR_MESSAGE()), PATINDEX(N'%, [0-9]%).', ERROR_MESSAGE()) - PATINDEX(N'%[0-9]%,%', ERROR_MESSAGE()))
```
I'm not changing the logic here, no reason to be concerned.

nmummau commented:
/review

nmummau commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/1_Schema.sql:
```diff
@@ -10,9 +10,9 @@ IF OBJECT_ID('__schema__.Streams', 'U') IS NULL
             StreamName NVARCHAR(850)      NOT NULL,
             [Version]  INT DEFAULT (-1)   NOT NULL,
-            CONSTRAINT PK_Streams PRIMARY KEY CLUSTERED (StreamId),
+            CONSTRAINT PK_Streams PRIMARY KEY CLUSTERED (StreamId) WITH (OPTIMIZE_FOR_SEQUENTIAL_KEY = ON),
```
@alexeyzimarev  are you OK with this being a breaking change and the SqlServerStore requiring SQL Server 2019 for this change? For me personally that seems like a reasonable requirement.

alexeyzimarev replied:
Yeah, I think it's ok.

nmummau commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/2_AppendEvents.sql:
```diff
@@ -1,50 +1,106 @@
-        INSERT INTO __schema__.Messages (MessageId, MessageType, StreamId, StreamPosition, JsonData, JsonMetadata, Created)
-        SELECT message_id, message_type, @stream_id, @current_version + (ROW_NUMBER() OVER(ORDER BY (SELECT NULL))), json_data, json_metadata, @created
-        FROM @messages
+        BEGIN TRANSACTION;
```
I'm not feeling like we don't need a transaction since the client wraps this entire call in a transaction.

nmummau replied:
transaction removed from the SQL code directly as the client is responsible for the transactions. This means issue #429  is a non-issue. Instead I added `SET XACT_ABORT ON;` to all procedures

nealmummau committed:
chore(mssql): enable XACT_ABORT for all stored procedures used by Eventuous

Added SET XACT_ABORT ON to all SQL stored procedures invoked through
GetStoredProcCommand() to ensure atomic rollback behavior when the client
transaction encounters an error. This prevents partial commits and
inconsistent transactional states during event append and stream operations.

Related to issue #429

nmummau commented:
/review

nealmummau committed:
fix(mssql): truncate_stream wasn't properly formatting and and sending the error message

nealmummau committed:
refactor(mssql): use SCOPE_IDENTITY() to get new StreamId rather than querying the table for the record we just inserted

alexeyzimarev commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/2_AppendEvents.sql:
```diff
@@ -1,50 +1,101 @@
+            message_id,
+            message_type,
+            @stream_id,
+            @current_version + CAST(ROW_NUMBER() OVER (ORDER BY (SELECT NULL)) AS INT),
```
Large batches aren't normal

alexeyzimarev commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/5_ReadStreamBackwards.sql:
```diff
@@ -4,24 +4,39 @@ CREATE OR ALTER PROCEDURE __schema__.read_stream_backwards
+    IF @current_version < @from_position + @count
+    BEGIN
+        RETURN;
+    END;
```
That is actually valid. It should be `@current_version < @from_position - @count`, I think.

nmummau replied:
For the record, this code that is being flagged by AI as incorrect is not a change I'm introducing. But I'm happy to fix it. I agree the current code is wrong. I'm actually writing tSQLt unit tests against these procedures for my project (future PR will add these to this repo too) and I have confirmed a fix to work well
This image outlines each of the options 
<img width="708" height="1437" alt="image" src="https://github.com/user-attachments/assets/d91e9a44-7906-4f5b-b80d-b0427a50e5da" />

alexeyzimarev replied:
Yeah, I know it's not your code. I flagged it so we don't forget to fix it.

alexeyzimarev commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/2_AppendEvents.sql:
```diff
@@ -1,50 +1,106 @@
+    SET XACT_ABORT ON;
+
+    IF NOT EXISTS (SELECT 1 FROM @messages)
+        THROW 50002, N'No messages to append', 1;
```
We probably could remove this check as the caller should ensure there are messages to append before calling the procedure.

nealmummau committed:
fix(mssql): correct backwards read bounds check

Replaced invalid guard `@current_version < @from_position + @count` with proper
range validation `@from_position < 0 OR @from_position > @current_version`.

Previous logic incorrectly returned no results when valid messages existed.
New check correctly validates start position independently of page size.

nealmummau committed:
fix(mssql): moved stream version update inside TRY block

and added `@@ROWCOUNT` check to detect concurrent updates.
If no rows are affected, the procedure now
throws a `WrongExpectedVersion` error, aligning behavior with duplicate
append handling.

- Prevents silent version overwrite on race conditions
- Ensures consistency with optimistic concurrency semantics
- Keeps existing WrongExpectedVersion error format for client handling

nmummau commented on src/SqlServer/src/Eventuous.SqlServer/Scripts/2_AppendEvents.sql:
```diff
@@ -1,50 +1,101 @@
+    SET [Version] = @new_version
+    FROM __schema__.Streams s
+    WHERE s.StreamId = @stream_id
+    AND s.[Version] = @current_version;
```
pushed new commit

nmummau commented:
/review
